Guard from cancellation that could bring down VS.

heejaechang opened the pull request:
created separate PR since there was some change around how cancellation works between 15.4 and 15.5 (Roslyn moved to new cancellation API supported by StreamJsonRpc)

**Customer scenario**

Users is shutting down VS, and see VS get crashed. and restarted.

**Bugs this fixes:**

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/476653?src=alerts&src-action=cta
https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/518997

**Workarounds, if any**

there is no workaround. it doesn't happen all the time though, it only happen if VS is shutting down at unfortunate time where there is on going active connection with OOP.

**Risk**

this is very safe fix. 

**Performance impact**

N/A

**Is this a regression from a previous update?**

No

**Root cause analysis:**

most of Roslyn is guarded against cancellation, and these are some places related to OOP that is not guarded on cancellation exception.

**How was the bug found?**

customer feedback, dogfooding

jasonmalinowski commented:
@heejaechang We'll port this automatically for you -- is there a reason to do this separately? This can otherwise cause merge conflicts for us.

heejaechang commented:
@jasonmalinowski code is slightly different since there was big change on OOP between 15.4 and 15.5. for example, in 15.5 we moved OOP to use new streamJson API which provide cancellation natively.

natidea commented:
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/518997

heejaechang commented:
@dotnet/roslyn-analysis @dotnet/roslyn-ide can I get some code review? simple change.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs:
```diff
@@ -84,13 +70,42 @@ public void OnSolutionRemoved()
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId),
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).Wait(CancellationToken.None);
+                }
+                catch (OperationCanceledException)
```
The caller passed CancellationToken.None, why are we getting cancellation? Shouldn't the error case just be a return of false since it's a Try* API? This seems like we're playing whack-a-mole here...

heejaechang replied:
I don't want the TryInvoke to blindly swallow all cancellation and return false. it will return false if it is one of our failures we defined that could happen such as users killed OOP process.

for this specific code path, only cancellation it can throw is if service hub failed (not us, but service hub itself). for those failure, we used to crash, but with long discussion, management decide not to crash but show info bar and log those issues and NFW with service hub log attached and move on.

how we move on is, creating a new cancellationToken and throw that on the fly.

you can see it here (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Remote/JsonRpcClient.cs,9eaa564ffdacf618)

comment seems describing old behavior, now we don't rethrow, but throw new cancellation token.

heejaechang replied:
tagging @jasonmalinowski

jasonmalinowski replied:
No, TryInvoke should not catch all exceptions. But it seems the expectation is every caller must do that so I fail to see the difference. :smile:

Why can't TryInvoke catch just that special cancellation token's cancellation and let the other pass through?

heejaechang replied:
because if TryInvoke catches it then, caller of TryInvoke need to handle it. but if it let cancellation pass through, one at the higher layer can handle cancellation same way it handles our own cancellation.

jasonmalinowski replied:
The expectation of .NET APIs is they don't throw OperationCancelledException with a token unrelated to the one they passed in. TPL and other APIs expect this -- this is why we regularly got bit by aggregate exceptions appearing when we screwed this up. If you pass CancellationToken.None, that means you don't expect cancellation. Yes, the call can always _fail_, but that is different than cancellation. If you want to throw a different exception and then catch that exception, I'm OK with that. I still can't explain why TryInvoke() is then bool-returning!

I think your PR here argues against yourself: the return of TryInvoke is already being ignored in various places, which means people are already ignoring the result. If you did the catch once, then you'd fix all of these.

heejaechang replied:
sure, I know how cancellation is supposed to work, but unfortunately, that is not the case here. we use cancellation token as a way to escape control flow and avoid crash from service hub. it would be nice to have other way, but right now, that is the way we settled for service hub failure. hopefully, once service hub issues are all fixed, we can remove this hack, but not in current form where we still have bunch of issues service hub is trying to fix.

...

"I still can't explain why TryInvoke() is then bool-returning!"

one of case is OOP process is not there. and not all call to service hub is TryInvoke. that is just one of way. I am not sure I am following you. in this particular one, it doesn't care whether it is failed or not since it is fire and forget. other caller do care and handle when TryInvoke failed.

heejaechang replied:
@jasonmalinowski I think people are confused by try  { call(CancellationToken.None) } catch (Cancellation)

I agree that is weird, but, like I said, this is the way we choose to handle service hub failure (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Remote/JsonRpcClient.cs,9eaa564ffdacf618). and ensure highest chance that will not bring VS down when that happen.

@jinujoseph I think I need you here. basically, @jasonmalinowski and @sharwell seems don't like the way we decide how to handle service hub failure to not crash VS.

this is done by after talking with service hub team, sri, matt to find highest chance not to crash VS. creating and throwing a cancellation token on the fly.

their object seems not on this code itself, but the fact it can throw its own cancellation token.

jasonmalinowski replied:
Indeed, this is most confusing, in a place where we need a 100% crystal clear API. The fact we're making a fix here indicates our own "let's not crash VS" pattern isn't working how we expected. Looking at the code doing the throwing:

https://github.com/dotnet/roslyn/blob/093cf36c56feefbf081b575378db41a6279ebe51/src/VisualStudio/Core/Next/Remote/JsonRpcClient.cs#L79-L81

...has a comment implying that was a temporary decision to guard against a problem, not something we expected to maintain long term. The PR comments imply the decision was a temporary workaround, not something we thought was good design.

heejaechang replied:
yep. and we still need that temporary solution. until these are fixed.

2: Bug 517974: VS crashes when %localappdata%\IdentityService\AccountStore.json throws UnauthorizedAccessException
10: Bug 517975: VS crashes when Identity service throws BadImageFormatException
38: Bug 518035: VsixServiceDiscovery randomly fails: ExternalSettingsManager::GetScopePaths failed to initialize PkgDefManager
Bug 518533: HubClient from devenv.exe calls into VS Installer's ServiceHub controller instead of VS's own

and more..

heejaechang replied:
> The fact we're making a fix here indicates our own "let's not crash VS" pattern isn't working how we expected

that is not true. it works most of time. since all the crashes watson we used to get got all disappeared. I have no idea where you got that conclusion.

this is only one report we got since we added the cancellation approach.

CyrusNajmabadi replied:
I don't understand this bit: "for this specific code path, only cancellation it can throw is if service hub failed (not us, but service hub itself)."

If that's how servicehub works, then we should be catching and handling that exception at the deepest level possible (i.e. when we actually call into servicehub, not when we're calling into our own abstraction around it).  Classes like KeepAliveSession should not leak otu this strange behavior.  But should instead wrap it and make it operate like a normal roslyn class with proper cancellation/exception semantics.

heejaechang commented:
@jinujoseph @jasonmalinowski well that is completely different discussion than this PR. this PR doesn't try to address whether we should throw our own cancellation or it should throw new custom exception.

we already decided to throw our own cancellation and it has been there since 15.3 or before. can't remember.

sure, we can discuss whether we should change our approach on how to service hub failure or not, but that's different discussion. 

and that will be highly risky change, so definitely not for 15.5. we choose to this cancellation approach since that removed most of crashes. if we introduce new exception or other approach, we need to make sure that doesn't increase chance of crashes when service hub failed.

jinujoseph commented:
Lets meet on monday to discuss the approach and clarify the concern

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs:
```diff
@@ -196,10 +196,10 @@ private void SetRemoteHostBitness()
-                    WellKnownRemoteHostServices.RemoteHostService, _workspace.CurrentSolution,
-                    nameof(IRemoteHostService.SynchronizeGlobalAssetsAsync),
-                    (object)checksums, cancellationToken).ConfigureAwait(false);
+                await client.IgnoreCancellationAsync(() =>
```
Seeing something like IgnoreCancellationAsync seems *extremely strange* to me.

1. either we don't want this to be cancellable, in which case, we probably shouldn't pass cancellation token.
2. we do want this to be cancellable.  But we want to swallow cancellation.  In which case, having an explicit try/catch for that condition seems more appropriate.

Regardless of which approach is taken (including keeping 'IgnoreCancellationAsync'), this sort of complexity really needs to be documented in code.  Code that does not follow normal patterns really needs to have appropriate explanations in it so that people maintaining it down the line know why it's operating in this manner.

CyrusNajmabadi commented:
What i'm taking away from reading all this is that we wrap certain subsystems that have 'interesting' behavior.  If that's the case, we should be pushing our special case handling of those subsystems to as low a level as possible.  Including, if necessary, just making simple wrappers around them that make them into well behaved objects in the roslyn space.  Then, we can use those wrappers from our higher level abstractions, needing no special code there.

RIght now we seem to be taking the approach of forcing the higher level layers to be aware of the strange behavior of the subsystems we depend on, adding very strange code in unclear ways.  There are several problems with this.  First off, it's very difficult to know if this change is sufficient.  Are there places we could have missed where we'll still have problems elsewhere?  Second, it's very hard to maintain.  Even if this code is correct, it's far too simple for someone to potentially introduces bugs later on.  For example, if someone else starts using a KeepAliveSession, will they know they have to have this sort of code?

These are leaky abstractions, and that's not something we want for such a critical subsystem.

heejaechang commented:
updated code based on our discussion in the meeting. realized I can create custom exception inherits from cancellation exception, so rather than having singleton we discussed, I went to custom exception path with cancellation exception as its super type.

added some more comments.

heejaechang commented:
@jinujoseph updated for 15.5. reported issue was for 15.4. so, I can port to 15.4, but its up to you.

jinujoseph commented:
lets get it reviewed for 15.5 , we can update the 15.4 PR later.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/JsonRpcClient.cs:
```diff
@@ -59,10 +59,9 @@ protected virtual void Disconnected(JsonRpcDisconnectedEventArgs e)
+                // See https://github.com/Microsoft/vs-streamjsonrpc/issues/67
                 cancellationToken.ThrowIfCancellationRequested();
 
                 LogError($"exception: {ex.ToString()}");
```
(having to comment here since GitHub doesn't let me comment outside the diff...)

Should the comment on line 69 ("this is to make us not crash...") be revised with a bit more detail? I admit my biggest confusion before was trying to decipher what was going on there.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/JsonRpcClient.cs:
```diff
@@ -134,10 +131,9 @@ protected virtual void Disconnected(JsonRpcDisconnectedEventArgs e)
-                // until we move to newly added cancellation support in JsonRpc, we will catch exception and translate to
-                // cancellation exception here. if any exception is thrown unrelated to cancellation, then we will rethrow
-                // the exception
+                // StreamJsonRpc throws RemoteInvocationException if the call is cancelled.
```
Can we extract this pattern (this throw, the log error, then the throw of our own cancellation) into a single function?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/JsonRpcClient.cs:
```diff
@@ -161,6 +157,10 @@ private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationTo
                 return true;
             }
 
+            // once we fix all pending vso bugs, we should remove these 
```
Can you clarify "pending VSO bugs" to make it clear this is relating to StreamJsonRpc or something? I presume you don't mean we fix _every_ bug in VSO.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs:
```diff
@@ -196,10 +196,18 @@ private void SetRemoteHostBitness()
+                        WellKnownRemoteHostServices.RemoteHostService, _workspace.CurrentSolution,
+                        nameof(IRemoteHostService.SynchronizeGlobalAssetsAsync), (object)checksums, cancellationToken).ConfigureAwait(false);
+                }
+                catch(RemoteHostClientExtensions.UnexpectedRemoteHostException)
```
I would have expected that TryRunRemoteAsync would be catching UnexpectedRemoteHostException and converting that to a false. Otherwise, you're effectively throwing an OperationCancelledException that wasn't from your original cancellation token. Yes, it's a derived exception but you get to follow the rules of your base type.

heejaechang replied:
this is special case for OOP engine itself. other ones, doesn't need to catch this exception. and TryRunRemoteAsync is API outside of OOP engine uses, so it doesn't catch this exception and let VS to soft crash.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs:
```diff
@@ -84,13 +70,44 @@ public void OnSolutionRemoved()
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId),
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).Wait(CancellationToken.None);
+                }
+                catch (RemoteHostClientExtensions.UnexpectedRemoteHostException)
```
Ditto: expected that this catch would have moved into TryInvokeAsync.

heejaechang replied:
same as above.

heejaechang commented:
I updated comments. I think one of confusion was people thinking the exception is something people should catch and recover. it is not. if the exception thrown, one should consider VS as crashed. like OOM happened.

tmat commented:
test windows_debug_vs-integration_prtest please

CyrusNajmabadi commented:
>  I think one of confusion was people thinking the exception is something people should catch and recover. it is not. 

...  But that's exactly what your code does.  It is a try/catch that completely swallows the exception.

CyrusNajmabadi commented:
> ignore unexpected remote host exception. it is allowed here since it is part of OOP engine.
> no one outside of engine should ever catch this exception or care about it.

If it's an internal detail of the engine, then why is it being thrown outside of that method?  And what does it mean for other code that calls into IRemoteHostClient.TryRunRemoteAsync? or KeepAliveSession.TryInvokeAsync?

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs:
```diff
@@ -84,13 +70,48 @@ public void OnSolutionRemoved()
+                catch (RemoteHostClientExtensions.UnexpectedRemoteHostException)
+                {
+                    // ignore unexpected remote host exception. it is allowed here since it is part of OOP engine.
+                    // no one outside of engine should ever catch this exception or care about it.
```
You say no one outside of engine shoudl ever catch this exception... but it looks like we're outside the engine, and catching it.  What constitutes "engine" in this description?

heejaechang replied:
see the filename. it is still inside of OOP engine.

jasonmalinowski replied:
I concur with @CyrusNajmabadi here: I can't actually tell what is considered internal and public and anything else here. If the exception type was some private class of something else I'd feel better that it's "internal" but right now somebody can take a dependency on this.

heejaechang replied:
due to layering that is not possible. I can make it internal, but I made it nested type of the extensions for that.

I still not getting that reasoning. unless all OOP types are moved to its own dll and make only those APIs that are supposed to be used by others are marked public, I am not sure how I understand your comment.

for example, ServiceHubRemoteHostClient is something nobody in Roslyn should ever create themselves. it is one of implementation of IRemoteHostClient. but since people can see everything in Roslyn, they can do it. and break whole system by creating another OOP process which lacks all internal states.

same goes for all other services, such as VisualStudioMetadataProvider and all which will break metadata identity, CompilationTracker which will break symbol's identity.

as Roslyn developer, it is their responsibility what is boundary of those internal types. it is cost of all Roslyn being internal visible to all its internal types and dlls.

CyrusNajmabadi replied:
> see the filename. it is still inside of OOP engine

What is the 'engine' in this design?  It's really not clear to me.  

> same goes for all other services, such as VisualStudioMetadataProvider

Our other services don't throw specific exceptions that are only supposed to be handled by some codepaths, but not by others. It's really hard to tell if this is right or not because there's no way to really know if you've gotten the entire list of callers that need protection.  Also, because the layers of the remote system aren't clear, it's not easy (or possible) for someone else to come in and make a determination if other locations are a problem.

heejaechang replied:
we are talking about internal types boundary. we are switching subject.

we know it is not clear. and that's fine since that was clearly communicated to the manager when this is added. consensus at the time was, as long as it make VS not hard crash, but soft crash most of time, and give users time to save their work, it is fine. it proved that did its work. whether it is going to work for 100%, no idea. and that's work for NFW change on Roslyn or moving back to crash behavior once most of issues we know got fixed.

until then, this is what we have. and this PR is fixing a few specific cases we know this workaround doesn't work. 

and the exception is exposed since it is inside of same OOP engine. it is written in comment again and again to indicate that. 

you need to read code to see layering. are there any complex features that have simple layering? see Options, Editor work for VS4Mac, editorconfig and etc.

heejaechang commented:
@dotnet/roslyn-analysis @dotnet/roslyn-ide ping? can someone take a look?

CyrusNajmabadi commented:
I'm still trying to figure out what this means:

> no one outside of engine should ever catch this exception or care about it.

So, if i'm using one of these services (like the KeepAliveSession), and this exception happens, it just tears down VS?  Why are some callers of KeepAliveSession guarded, but others are not?  It KeepAliveSession only supposed to be used by certain parts of hte system?  If so, why do we use it for things like the Nuget indexing service?

heejaechang commented:
@CyrusNajmabadi 
> So, if i'm using one of these services (like the KeepAliveSession), and this exception happens, it just tears down VS? Why are some callers of KeepAliveSession guarded, but others are not? It KeepAliveSession only supposed to be used by certain parts of hte system? If so, why do we use it for things like the Nuget indexing service?

who guards on it? KeepAliveSession doesn't. for users who write things like FAR, Navigate To, if this exception thrown, then it should be considered VS is crashed. and they should ever know or care about this exception.

CyrusNajmabadi commented:
> we know it is not clear. and that's fine since that was clearly communicated to the manager when this is added. 

The problem is this: how can we review PRs with confidence when things are confusing and unclear?  How do we know that this doesn't make things worse?

What i'm trying to propose is an approach whereby we isolate any of the root problems (which appear to be servicehub oddities) at as low a level as possible.  We then only expose types to the outer system that have well behaving, roslyn-like, semantics.  Then, we can feel more confident that things will work as we expect and that the disparity between systems is isolated.

CyrusNajmabadi commented:
> who guards on it? 

You have code in this review that guards on it.  For example, the code in RegisterPrimarySolution.  This code specifically used KeepAliveSession and swallows and catches this specific exception type.  It's not clear why this code needs to do that, but other code that uses KeepAliveSession does not.  Furthermore, let's give a simple example.  Say you went on vacation and someone needed to do some work here.  Maybe they needed to make another call to TryInvokeAsync.  Would they need to catch this exception or not.  From this review, i don't see any documentation on TryInvokeAsync that lets any callers know what they should be doing here.  But it seems like it's critical that "the engine" (which is not really defined anywhere afaict) catch this exception in some cases.  How would someone know that?  How would this be caught in a code review if they didn't do that?  etc. etc.

CyrusNajmabadi commented:
I think there needs to be clear documentation explaining things like "outside of OOP engine".  What constitutes the "oop engine"?  What is the "inside" part of it, and what is the "outside" part of it?  Method that throw, but are expected to be handled differently based on teh caller should also likely be extensively documented explaining their semantics and how they should be used properly.   

Finally, for code that behaves differently than other pieces of code (for example, the usage of KeepAliveSession in the nuget caching service) we should likely *proactively* add comments explaining the situation to make things clear for the next person who comes along.  For example, if someone needs to look at this a year from now, they may go "why is there no try/cathc over here when all the rest of this code has it."  Sometimes we need to comment the *absence* of code to make sure things are clear moving forward.

heejaechang commented:
I talked to @jinujoseph, she will take it from here. I don't know how to say more on this exception doesn't leak from OOP engine.

CyrusNajmabadi commented:
Your latest commit (https://github.com/dotnet/roslyn/pull/22968/commits/1ceea420b9951d6d45028ba353f427edfbab51c0) is a good example of the concern i have around this approach.  In that latest commit you made this change:

```
-            private Task SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
 +            private async Task SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
              {
 -                return _service.Workspace.SynchronizePrimaryWorkspaceAsync(_service.Workspace.CurrentSolution, cancellationToken);
 +                try
 +                {
 +                    await _service.Workspace.SynchronizePrimaryWorkspaceAsync(_service.Workspace.CurrentSolution, cancellationToken).ConfigureAwait(false);
 +                }
 +                catch (JsonRpcEx.UnexpectedRemoteHostException)
 +                {
 +                    // ignore unexpected remote host exception. it is allowed here since it is part of OOP engine.
 +                    // no one outside of engine should ever catch this exception or care about it.
 +                    // we catch here so that we don't physically crash VS and give users time to save and exist VS
 +                }
              }
```

i.e. here was a place that was missed from the previous commits that we needed this new try/catch added to.  This was missed by all reviewers.  And, as i mentioned earlier, there is no way for a reviewer (like myself) to know if the current work is complete, or if there is more stuff missing.  Furthermore, as changes happen here in the future, it's still unclear to me how someone is supposed to know if/when/how they should follow this pattern.

Given that this was missed *in this very PR*, i find this approach pretty darn concerning.  :-/

heejaechang commented:
@CyrusNajmabadi it wasn't missed. I made catch inside of _service.Workspace.SynchronizePrimaryWorkspaceAsync itself. and you were complaining about layering, so I moved it up to caller of it.

jinujoseph commented:
@nat @tmat   this is not targeted for 15.5 any more

heejaechang commented:
changed to dev15.6 per discussion with @jinujoseph

heejaechang commented:
ping? I would like to check this in for 15.6

heejaechang commented:
retest this please

heejaechang commented:
ping? @tmat just had hit this issue. this should guard us from hard crashing VS in crash situation.

heejaechang commented:
ping? @dotnet/roslyn-ide @dotnet/roslyn-analysis

heejaechang commented:
@mavasani @sharwell @jasonmalinowski can you take a look? I would like to check this in.

heejaechang commented:
closing this. will have another PR that formalize safe crash concept a bit more.
